File: baekjoon/basic2/1068/1068.cpp
// 백준 1068 : 트리 (https://www.acmicpc.net/problem/1068)
#include <iostream>
#include <vector>
using namespace std;
int n, parents, target, cnt;
vector<int> adj[51], root;
int del[51], visited[51];

void off(int u){
    del[u] = 1;
    for(int v : adj[u]){
        if(del[v]) continue;
        off(v);
    }
    return;
}

void dfs(int u){
    visited[u] = 1;
    bool is_leaf = true;
    for(int v : adj[u]){
        if(visited[v] || del[v]) continue;
        dfs(v);
        is_leaf = false;
    }
    if(is_leaf) cnt++;
    return;
}

int main(){
    ios_base::sync_with_stdio(false);
    cin.tie(NULL); cout.tie(NULL);

    cin >> n;
    for(int i=0; i<n; i++){
        cin >> parents;
        if(parents == -1){
            root.push_back(i);
            continue;
        }
        adj[parents].push_back(i);
    }
    cin >> target;
    off(target);
    for(int u : root){
        if(del[u]) continue;
        dfs(u);
    }

    cout << cnt << "\n";

    return 0;
}

File: baekjoon/basic2/1068/1068.py
# 백준 1068 : 트리 (https://www.acmicpc.net/problem/1068)
import sys 
sys.setrecursionlimit(10**6)

def dfs(u):
    visited[u] = 1
    ret = 0
    child = 0
    for v in adj[u]:
        if v == target or visited[v]:
            continue
        child += 1
        ret += dfs(v)
    if child == 0:
        ret += 1
    return ret

n = int(sys.stdin.readline())

adj = [[] for _ in range(n)]
visited = [0 for _ in range(n)]

node = list(map(int, sys.stdin.readline().split()))

for i in range(n):
    if node[i] == -1:
        root = i
    else:
        adj[node[i]].append(i)

target = int(sys.stdin.readline())

if target == root:
    sys.stdout.write("0\n")
else:
    ret = dfs(root)
    sys.stdout.write(str(ret)+"\n")